File: admin-backend/app/models/theater.py
"""
智能剧场模型 - Phase 3: 多账号协同表演
"""

from datetime import datetime
from typing import Optional, List, Dict, Any
import uuid
from sqlalchemy import Column, String, Integer, Boolean, JSON, DateTime, Text
from sqlalchemy.sql import func

from app.db import Base


def generate_uuid():
    """生成UUID字符串"""
    return str(uuid.uuid4())


class TheaterScenario(Base):
    """剧场场景表"""
    __tablename__ = "theater_scenarios"
    
    id = Column(String(36), primary_key=True, default=generate_uuid)
    name = Column(String(200), nullable=False, index=True)  # 场景名称
    description = Column(Text, nullable=True)  # 场景描述
    roles = Column(JSON, nullable=False, default=list)  # 角色列表: ["UserA", "UserB", "UserC"]
    timeline = Column(JSON, nullable=False, default=list)  # 时间轴动作列表
    # timeline 结构示例:
    # [
    #   { "time_offset": 2, "role": "UserA", "content": "看多", "action": "send_message" },
    #   { "time_offset": 5, "role": "UserB", "content": "看空", "action": "send_message" },
    #   { "time_offset": 10, "role": "UserA", "content": "我也看空", "action": "send_message" }
    # ]
    enabled = Column(Boolean, default=True, nullable=False, index=True)  # 是否启用
    created_by = Column(String(100), nullable=True)  # 创建者用户ID
    created_at = Column(DateTime, default=func.now(), nullable=False)
    updated_at = Column(DateTime, default=func.now(), onupdate=func.now(), nullable=False)


class TheaterExecution(Base):
    """剧场执行记录表"""
    __tablename__ = "theater_executions"
    
    id = Column(String(36), primary_key=True, default=generate_uuid)
    scenario_id = Column(String(36), nullable=False, index=True)  # 关联的场景ID
    scenario_name = Column(String(200), nullable=False)  # 场景名称（快照）
    group_id = Column(String(100), nullable=False, index=True)  # 目标群组ID
    agent_mapping = Column(JSON, nullable=False, default=dict)  # Agent 映射: {"UserA": "agent_id_1", "UserB": "agent_id_2"}
    status = Column(String(20), nullable=False, index=True, default="pending")  # 状态: pending, running, completed, failed, cancelled
    started_at = Column(DateTime, nullable=True)  # 开始时间
    completed_at = Column(DateTime, nullable=True)  # 完成时间
    duration_seconds = Column(Integer, nullable=True)  # 执行时长（秒）
    executed_actions = Column(JSON, default=list)  # 已执行的动作列表（用于重试和恢复）
    error_message = Column(Text, nullable=True)  # 错误信息
    created_by = Column(String(100), nullable=True)  # 创建者用户ID
    created_at = Column(DateTime, default=func.now(), nullable=False, index=True)
    updated_at = Column(DateTime, default=func.now(), onupdate=func.now(), nullable=False)

File: saas-demo/src/lib/api/theater.ts
/**
 * 智能剧场 API 客户端
 */

import { fetchWithAuth } from "./client"
import { getApiBaseUrl } from "./config"

const API_BASE = getApiBaseUrl()

export interface TimelineAction {
  time_offset: number
  role: string
  content: string
  action?: string
  payload?: Record<string, any>
}

export interface TheaterScenario {
  id: string
  name: string
  description?: string
  roles: string[]
  timeline: TimelineAction[]
  enabled: boolean
  created_by?: string
  created_at: string
  updated_at: string
}

export interface TheaterScenarioCreate {
  name: string
  description?: string
  roles: string[]
  timeline: TimelineAction[]
  enabled?: boolean
}

export interface TheaterScenarioUpdate {
  name?: string
  description?: string
  roles?: string[]
  timeline?: TimelineAction[]
  enabled?: boolean
}

export interface TheaterExecutionCreate {
  scenario_id: string
  group_id: string
  agent_mapping: Record<string, string>  // { "UserA": "agent_id_1", "UserB": "agent_id_2" }
}

export interface TheaterExecution {
  id: string
  scenario_id: string
  scenario_name: string
  group_id: string
  agent_mapping: Record<string, string>
  status: "pending" | "running" | "completed" | "failed" | "cancelled"
  started_at?: string
  completed_at?: string
  duration_seconds?: number
  executed_actions: any[]
  error_message?: string
  created_by?: string
  created_at: string
  updated_at: string
}

export interface TheaterScenarioListParams {
  skip?: number
  limit?: number
  enabled?: boolean
}

export interface TheaterExecutionListParams {
  skip?: number
  limit?: number
  status?: string
}

/**
 * 获取场景列表
 */
export async function getScenarios(params?: TheaterScenarioListParams): Promise<TheaterScenario[]> {
  const queryParams = new URLSearchParams()
  if (params?.skip !== undefined) queryParams.append("skip", params.skip.toString())
  if (params?.limit !== undefined) queryParams.append("limit", params.limit.toString())
  if (params?.enabled !== undefined) queryParams.append("enabled", params.enabled.toString())
  
  const url = `${API_BASE}/theater/scenarios${queryParams.toString() ? `?${queryParams.toString()}` : ""}`
  const response = await fetchWithAuth(url)
  
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`)
  }
  
  return response.json()
}

/**
 * 获取场景详情
 */
export async function getScenario(scenarioId: string): Promise<TheaterScenario> {
  const response = await fetchWithAuth(`${API_BASE}/theater/scenarios/${scenarioId}`)
  
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`)
  }
  
  return response.json()
}

/**
 * 创建场景
 */
export async function createScenario(data: TheaterScenarioCreate): Promise<TheaterScenario> {
  const response = await fetchWithAuth(`${API_BASE}/theater/scenarios`, {
    method: "POST",
    body: JSON.stringify(data),
  })
  
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: `HTTP ${response.status}` }))
    throw new Error(error.detail || `HTTP ${response.status}`)
  }
  
  return response.json()
}

/**
 * 更新场景
 */
export async function updateScenario(
  scenarioId: string,
  data: TheaterScenarioUpdate
): Promise<TheaterScenario> {
  const response = await fetchWithAuth(`${API_BASE}/theater/scenarios/${scenarioId}`, {
    method: "PUT",
    body: JSON.stringify(data),
  })
  
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: `HTTP ${response.status}` }))
    throw new Error(error.detail || `HTTP ${response.status}`)
  }
  
  return response.json()
}

/**
 * 删除场景
 */
export async function deleteScenario(scenarioId: string): Promise<void> {
  const response = await fetchWithAuth(`${API_BASE}/theater/scenarios/${scenarioId}`, {
    method: "DELETE",
  })
  
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: `HTTP ${response.status}` }))
    throw new Error(error.detail || `HTTP ${response.status}`)
  }
}

/**
 * 执行场景
 */
export async function executeScenario(data: TheaterExecutionCreate): Promise<TheaterExecution> {
  const response = await fetchWithAuth(`${API_BASE}/theater/execute`, {
    method: "POST",
    body: JSON.stringify(data),
  })
  
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: `HTTP ${response.status}` }))
    throw new Error(error.detail || `HTTP ${response.status}`)
  }
  
  return response.json()
}

/**
 * 获取执行记录列表
 */
export async function getExecutions(params?: TheaterExecutionListParams): Promise<TheaterExecution[]> {
  const queryParams = new URLSearchParams()
  if (params?.skip !== undefined) queryParams.append("skip", params.skip.toString())
  if (params?.limit !== undefined) queryParams.append("limit", params.limit.toString())
  if (params?.status) queryParams.append("status", params.status)
  
  const url = `${API_BASE}/theater/executions${queryParams.toString() ? `?${queryParams.toString()}` : ""}`
  const response = await fetchWithAuth(url)
  
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`)
  }
  
  return response.json()
}

/**
 * 取消执行
 */
export async function cancelExecution(executionId: string): Promise<void> {
  const response = await fetchWithAuth(`${API_BASE}/theater/executions/${executionId}/cancel`, {
    method: "POST",
  })
  
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: `HTTP ${response.status}` }))
    throw new Error(error.detail || `HTTP ${response.status}`)
  }
}
